docs/github.md

```diff
@@ -0,0 +1,107 @@
+# Using Jujutsu with GitHub Projects
+
+This guide assumes a basic understanding of either Git or Mercurial. 
+
+## Setup an SSH key
+As of December 2022 it's recommended to set up a ssh key to work with Github
+projects. See [Github's Tutorial][gh]. This restriction may be lifted in the 
+future, see [issue #469][http-auth] for more information and progress on authenticated http. 
+
+
+## Basic workflows
+You can either create a branch and then commit revisions on top of it (similiar 
+to how Git is typically used) or just start making changes in the repository
+(which is how Mercurial is used). 
+This allows you to continue using your favorite way of using the VCS. 
+
+It's also possible to just commit multiple changes and set the expected branch 
+retroactively. 
+
+**Notable Differences to other VCS**  
+Branches do not move forward when creating commits on top. This means you should
+set the current branch to work on before starting your work. 
```

Comment:
This is confusing to me. I would say that you should create the branch after finishing your work instead, since you'll have to do that anyway, so setting a branch before starting is just unnecessary. I suspect you're thinking of the case of single-commit PR, but I don't want users to get used to that case and then be confused when they create more than one commit.

Reply:
Both are confusing to me... 
What is "move forward" referring to a branch? What is "on top"?

Creating a branch after you work seems pointless. In my mind, branches keep your work-in-progress separate from your stable code, so you should branch first, test, debug, and commit when it's working. Only after all that should the branch be merged back to stable.

Reply:
@martinvonz Is the command sequence above what you do on this repo? I find this repo so hard to follow because there are pushes that eliminate previous ones, and they all have the same commit messages. I hope that's not what you are promoting.

Reply:
> Both are confusing to me... What is "move forward" referring to a branch? What is "on top"?

The both mean "up" in the direction that `jj/git/hg log` and `gitk` show the graph. In Git, if you check out a branch (e.g. `git checkout -b my-branch`) and create a commit, then `my-branch` will point to the new commit.

> 
> Creating a branch after you work seems pointless. In my mind, branches keep your work-in-progress separate from your stable code, so you should branch first, test, debug, and commit when it's working. Only after all that should the branch be merged back to stable.

I think this confusion is all about "named branches" vs. "anonymous branches". See https://www.mercurial-scm.org/wiki/BranchingExplained. I think both @PhilipMetzger and I were talking about branches as the associated name, not the topological concept. We really need a document describing anonymous branches... (we can probably copy a lot from that Mercurial page).

Reply:
OK, it doesn't help future readers to explain the words to me here...it has to be in the doc itself.

Reply:
> @martinvonz Is the command sequence above what you do on this repo?

Yes, that's pretty typical for me to do in this repo (and other repos I occasionally contribute to). I mean, creating commits and pushing them to a remote is pretty much how everyone interacts with any repo :)

> I find this repo so hard to follow because there are pushes that eliminate previous ones, and they all have the same commit messages.

We use what's often called "stacked diffs". Googling that led me to https://jg.gg/2018/09/29/stacked-diffs-versus-pull-requests/. So consider reading that if you're unfamiliar with that way of working. FWIW, the Git and Mercurial projects also do that. I think most developers at Google and Meta also do.

Unfortunately, GitHub has pretty poor support for stacked diffs :( There are tools like graphite.dev and reviewable.io, which I think both attempt to have better support for stacked diffs.

> I hope that's not what you are promoting.

I definitely do promote stacked diffs :)

Oh, in case you were thinking that this document (or my suggested snippet?) somehow promoted stacked diffs, I don't think it does. I had planned to write about both workflows here (that was part of the draft I lost on reboot).

Reply:
> We use what's often called "stacked diffs".
> I definitely do promote stacked diffs

I read the link and a lot of links from there also. I understand the concept, however, it is still very hard to follow this repo because clicking links in the notifications often gets a 404 and the commit messages are the same in each push. So if you are simply using `jj` on this repo, and describing that in this doc, you really need to state that and link to the explanations in the doc, as well as help mitigate the confusion of new commits with same message.

Reply:
Good point, @joyously. I had a section (in that comment I lost to a reboot) describing the two different workflows for addressing PR comments, depending on whether the reader's project uses stacked diffs (i.e. clean commits and force-pushing) or the usual commit-on-top workflow. I've tried to recreate , now also clarifying that this project prefers clean commits. @PhilipMetzger, can you incorporate the below (or something similar to it)?


## Addressing review comments

There are two different workflows for addressing review comments, depending on your project's preference. Many projects prefer that you address review comments by adding new commits to your branch. Some projects (such as this one) instead prefer that you keep your commits clean by rewriting them and then force-pushing [^stacked-diffs].

[^stacked-diffs]: If you're wondering why we prefer clean commits in this project, see e.g. [this blog post](https://jg.gg/2018/09/29/stacked-diffs-versus-pull-requests/).

### Adding new commits

If your project prefers that you address review comments by adding commits on top, you can do that by doing something like this:

```shell
# Create a new commit on top of the `bar` branch from above. (Feel free to rename the branch, @PhilipMetzger)
$ jj new bar
# <Address the comment by updating the code>
# Review the changes
$ jj diff
# Give the fix description and create a new working-copy commit on top
$ jj commit -m 'address a comment'
# Update the branch to point to the new commit
$ jj branch set bar -r @-
# Push it to the remote
$ jj git push
```

### Rewriting commits

If your project prefers that you keep commits clean, you can do that by doing something like this:

```
# Create a new commit on top of the second-to-last commit in
#`bar` (let's say that's where we were asked to fix something)
$ jj new bar-
# <Address the comment by updating the code>
# Review the changes
$ jj diff
# Squash the changes into the parent commit
$ jj squash
# Push the updated branch to the remote. This will automatically become a force push
$ jj git push --branch bar
```

Reply:
I see that the two workflows look very similar. Can the one with rewrite change the message also? (I know it's replacing the original, so the message _should_ be mostly the same, but it did change...) Also, can you mention that in general, force pushing to a remote will cause problems for everyone else (404s in GitHub, contributors having to rework something their change might be based on)?

Reply:
I can suggest that. But it's actually not a workflow failure, but an issue how GitHub handles it in their backend. Because both [Phabricator](https://phabricator.com) and [Gerrit](https://www.gerritcodereview.com) are able to differentiate between multiple pushes with the same commit message.

Reply:
Regardless, this doc is about GitHub, so I think you should mention it.

Reply:
I added this between the squash and push step. 
```bash
$ # Edit the commit message to make your changes visible to your reviewers.
$ jj describe
```

Reply:
> I added this between the squash and push step.
> 
> ```shell
> $ # Edit the commit message to make your changes visible to your reviewers.
> $ jj describe
> ```

I'd prefer to remove this part again. It's quite atypical that review comments lead to an updated commit message. Of course it happens sometimes, but it's much less than than 50% of the time in my experience (even less than 10%?).

Reply:
> Also, can you mention that in general, force pushing to a remote will cause problems for everyone else (404s in GitHub, contributors having to rework something their change might be based on)?

Note that even GitHub has its "Compare" button that shows changes compared to before a force-push. That doesn't always work (if you're unlucky, the old commit could have been GC'd), and it will also include changes caused by rebasing (I hope they'll some day show output like `jj interdiff` instead).

I also think it's out of scope for this page since we already tell the user to do whatever their project prefers. I think the reader typically doesn't have control over which workflow their project uses (they do if it's their own personal project, but then they would typically not use PRs). Still, I agree that it's fair to add a footnote about the disadvantages, just like we have a footnote about the advantages (though I added that footnote only because you pointed out that readers might be confused about our decision). Maybe it's better to add a pointer to an external page about the pros and cons instead (i.e. a less opinionated page than the one I linked to).

Reply:
> I also think it's out of scope for this page since we already tell the user to do whatever their project prefers. I think the reader typically doesn't have control over which workflow their project uses (they do if it's their own personal project, but then they would typically not use PRs). Still, I agree that it's fair to add a footnote about the disadvantages, just like we have a footnote about the advantages (though I added that footnote only because you pointed out that readers might be confused about our decision).

I agree. Also I think we should leave our footnote, as it's preferred in this project and adds a small explanation.